concent_api/core/payments/bankster.py

```diff
@@ -264,80 +235,115 @@ def settle_overdue_acceptances(
     assert len(provider_ethereum_address) == ETHEREUM_ADDRESS_LENGTH
     assert provider_ethereum_address != requestor_ethereum_address
 
+    with transaction.atomic(using='control'):
+        requestor_client = Client.objects.get_or_create_full_clean(
+            public_key=requestor_public_key,
+        )
```

Comment:
If it is already present in the database, it should be returned without error.

Reply:
Not if it gets created between the `get` and the `create`. It's the race condition you always get when doing a non-atomic get or create. Same problem as in https://github.com/golemfactory/concent/issues/486

Reply:
I don't understand how it should be repeated then? This method should be called again in `except` block?

Reply:
The block should be in a context manager. If an exception with the right postgresql error code is raised, the manager should run the block again.

Or you could put it in a local function, run it in a `try ... except` block and run it again if you get the right exception.

The idea is simply that if you try to create it and the creation fails because the object exists you should try to fetch it instead. The current code tries to fetch first and only then creates and this introduces a race condition. Note that the create may not actually fail and you may get an error when the transaction is committed instead. Retrying once fixes the problem but we don't want to rerun the whole view. Just the part that creates the objects. It's what @pawelkisielewicz has been fixing with regard to subtasks lately. Since that's done, I'd prefer to avoid introducing this race condition in any new code.

Reply:
:+1: done

Reply:
Since there are two separate "get or create" operations and each one should be in a block of its own. So you need two blocks.

Also, I think catching just `IntegrityError` is too broad. You should check the error code from postgresql. See https://github.com/golemfactory/concent/issues/759#issuecomment-417295320

BTW, just duplicating it is fine for now but I think we should have a context manager for this because this case will come up very often (I've already found it in the other pull request).